File: y2018/python/day6/day6.py
# DISCLAIMER: really slow, probably a better way but idk

P2_MAX_DIST = 10 ** 4

points = []
min_x = float('inf')
max_x = 0
min_y = float('inf')
max_y = 0
with open('day6.txt') as read:
    for p in read.readlines():
        x, y = [int(i) for i in p.split(',')]
        assert x >= 0 and y >= 0

        min_x = min(min_x, x)
        max_x = max(max_x, x)
        min_y = min(min_y, y)
        max_y = max(max_y, y)

        points.append((x, y))

area = [[-1 for _ in range(max_y + 1)] for _ in range(max_x + 1)]
occ_amt = [0 for _ in range(len(points))]
for x in range(max_x + 1):
    for y in range(max_y + 1):
        min_dist = float('inf')

        distances = []
        for p in points:
            distances.append(abs(p[0] - x) + abs(p[1] - y))

        closest = min(distances)
        if distances.count(closest) == 1:
            area[x][y] = distances.index(closest)

        if area[x][y] != -1:
            occ_amt[area[x][y]] += 1

# -1 might be in this set but it doesn't matter
inf_area = set()
for x in range(min_x, max_x + 1):
    inf_area.add(area[x][min_y])
    inf_area.add(area[x][max_y])
for y in range(min_y, max_y + 1):
    inf_area.add(area[min_x][y])
    inf_area.add(area[max_x][y])

max_area = 0
for v, p in enumerate(points):
    if v not in inf_area:
        max_area = max(max_area, occ_amt[area[p[0]][p[1]]])

print(f"largest area: {max_area}")

close_enough_amt = 0
check_dist = P2_MAX_DIST // len(points) + 1
for x in range(min_x - check_dist, max_x + check_dist):
    for y in range(min_y - check_dist, max_y + check_dist):
        total_dist = sum(
            abs(p[0] - x) + abs(p[1] - y) for p in points
        )
        close_enough_amt += total_dist < P2_MAX_DIST

print(f"area of \"close enough\" region: {close_enough_amt}")

File: y2018/rust/src/bin/day6.rs
use std::collections::{HashMap, HashSet};
use std::fs;
use num::abs;

const P2_MAX_DIST: u32 = 1e4 as u32;

fn main() {
    let mut min_x = i32::MAX;
    let mut max_x = i32::MIN;
    let mut min_y = i32::MAX;
    let mut max_y = i32::MIN;
    // yk this definitely takes more lines than a normal for loop
    let points: HashSet<(i32, i32)> = fs::read_to_string("../input/day6.txt")
        .unwrap()
        .lines()
        .map(|p| {
            let split: Vec<&str> = p.split(',').map(|s| s.trim()).collect();
            let x = split[0].parse::<i32>().unwrap();
            let y = split[1].parse::<i32>().unwrap();
            min_x = min_x.min(x);
            max_x = max_x.max(x);
            min_y = min_y.min(y);
            max_y = max_y.max(y);
            (x, y)
        })
        .collect();

    let mut inf_pts = HashSet::new();
    let mut point_areas: HashMap<(i32, i32), u32> = points.iter().map(|p| (*p, 0u32)).collect();
    for x in min_x..=max_x {
        for y in min_y..=max_y {
            let mut smallest_dist = u32::MAX;
            let mut valid_pts = Vec::new();
            for p in &points {
                let dist = (abs(p.0 - x) + abs(p.1 - y)) as u32;
                if dist < smallest_dist {
                    smallest_dist = dist;
                    valid_pts = vec![p];
                } else if dist == smallest_dist {
                    valid_pts.push(p);
                }
            }

            if valid_pts.len() == 1 {
                let pt = valid_pts[0];
                *point_areas.get_mut(&pt).unwrap() += 1;
                if x == min_x || x == max_x || y == min_y || y == max_y {
                    inf_pts.insert(pt.clone());
                }
            }
        }
    }

    let largest_area = point_areas
        .iter()
        .filter(|(p, _)| !inf_pts.contains(p))
        .max_by(|(_, a1), (_, a2)| a1.cmp(a2))
        .unwrap()
        .1;

    println!("largest non-infinite area: {largest_area}");

    let mut close_enough_amt = 0;
    let check_dist = (P2_MAX_DIST / points.len() as u32 + 1) as i32;
    for x in (min_x - check_dist)..=(max_x + check_dist) {
        for y in (min_y - check_dist)..=(max_y + check_dist) {
            let total_dist: u32 = points
                .iter()
                .map(|p| (abs(p.0 - x) + abs(p.1 - y)) as u32)
                .sum();
            close_enough_amt += (total_dist <  P2_MAX_DIST) as u32;
        }
    }

    println!("area of \"close enough\" region: {close_enough_amt}");
}
